Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,199 @@
+        /// </summary>
+        private const int DefaultNumberOfBuckets = 20;
+
+        /// <summary>
+        /// The default number of entries stored in each <see cref="Bucket"/> instance.
+        /// </summary>
+        private const int DefaultEntriesPerBucket = 64;
+
+        /// <summary>
+        /// The current array of <see cref="Bucket"/> instances in use.
+        /// </summary>
+        private readonly Bucket[] buckets;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StringPool"/> class.
+        /// </summary>
+        public StringPool()
+            : this(DefaultNumberOfBuckets, DefaultEntriesPerBucket)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StringPool"/> class.
+        /// </summary>
+        /// <param name="numberOfBuckets">The total number of buckets to use.</param>
+        /// <param name="entriesPerBucket">The maximum number of <see cref="string"/> entries per bucket.</param>
+        public StringPool(int numberOfBuckets, int entriesPerBucket)
+        {
+            if (numberOfBuckets <= 0)
+            {
+                ThrowArgumentOutOfRangeException(nameof(numberOfBuckets));
+            }
+
+            if (entriesPerBucket <= 0)
+            {
+                ThrowArgumentOutOfRangeException(nameof(entriesPerBucket));
+            }
+
+            Span<Bucket> span = this.buckets = new Bucket[numberOfBuckets];
+
+            // We preallocate the buckets in advance, since each bucket only contains the
+            // array field, which is not preinitialized, so the allocations are minimal.
+            // This lets us lock on each individual buckets when retrieving a string instance.
+            foreach (ref Bucket bucket in span)
+            {
+                bucket = new Bucket();
+            }
+
+            NumberOfBuckets = numberOfBuckets;
+            EntriesPerBucket = entriesPerBucket;
+        }
+
+        /// <summary>
+        /// Gets the default <see cref="StringPool"/> instance.
+        /// </summary>
+        public static StringPool Default { get; } = new StringPool();
+
+        /// <summary>
+        /// Gets the total number of buckets in use.
+        /// </summary>
+        public int NumberOfBuckets { get; }
+
+        /// <summary>
+        /// Gets the maximum number of <see cref="string"/> entries stored in each bucket.
+        /// </summary>
+        public int EntriesPerBucket { get; }
+
+        /// <summary>
+        /// Gets a cached <see cref="string"/> instance matching the input content, or creates a new one.
+        /// </summary>
+        /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+        /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+        public string GetOrAdd(ReadOnlySpan<char> span)
```

Comment:
Should also be a `string` overload, to elide the copy in cases it isn't necessary

Reply:
Why do you need a `string` overload given that there's an implicit conversion to `ReadOnlySpan<char>` anyway? 🤔

Reply:
Because the ROS API copies, whereas a `string` one wouldn't need to

Reply:
Added a new `Add(string)` method in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3380/commits/194e082f118db738af60fe16966dd1472b071dbe, thank you for the suggestion! 🚀